sklearn/tree/_criterion.pyx

```diff
@@ -446,42 +435,39 @@ cdef class ClassificationCriterion(Criterion):
                 if sample_weight != NULL:
                     w = sample_weight[i]
 
-                for k in range(n_outputs):
-                    label_index = (k * label_count_stride +
-                                   <SIZE_t> y[i * y_stride + k])
-                    label_count_left[label_index] += w
+                for k in range(self.n_outputs):
+                    label_index = (k * self.sum_stride +
+                                   <SIZE_t> y[i * self.y_stride + k])
+                    sum_left[label_index] += w
 
-                diff_w += w
+                self.weighted_n_left += w
+                self.weighted_n_right -= w
```

Comment:
Why not putting later on `self.weighted_n_right = self.weight_n_total - self.weighted_n_left`. This remove the diff_w stuff, but keep the optim. What do you think? It will be also a bit clearer.